contract/src/main.rs

```diff
@@ -989,6 +1068,29 @@ pub extern "C" fn transfer() {
         Option::<Key>::None,
     );
 
+    let events_mode: u8 = get_stored_value_with_user_errors(
+        crate::constants::EVENTS_MODE,
+        NFTCoreError::MissingEventMode,
+        NFTCoreError::InvalidEventMode,
+    );
+
+    if events_mode != 0 {
+        events::record_event(match events_mode {
+            3 => Event::Cep47Dict(CEP47Event::Transfer {
+                sender: token_owner_key,
+                recipient: target_owner_key,
+                token_id: token_identifier.clone(),
+            }),
+            2 => Event::Cep47(CEP47Event::Transfer {
+                sender: token_owner_key,
+                recipient: target_owner_key,
+                token_id: token_identifier.clone(),
+            }),
+            1 => Event::Cep78,
+            _ => revert(NFTCoreError::InvalidEventMode),
+        });
+    }
+
     if let OwnerReverseLookupMode::Complete = utils::get_reporting_mode() {
```

Comment:
We are not emitting events, we are recording them. These changes only actually happen if the execution happens without any reverts or failures. In case the code reverts somewhere, no change in the global state is stored.